enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md

```diff
@@ -0,0 +1,503 @@
+
+Will work by annotating the ServiceAccount triggering the projection of the service account tokens into pods created for
+the operator (same method and resultant actions as: https://github.com/openshift/hypershift/pull/1351) the README here 
+gives an AWS specific example: https://github.com/openshift/aws-pod-identity-webhook)
+
+Possibly some changes to logic for Pod Identity Webhooks as needed for fine-grained credential
+management to allow for multi-tenancy (see definition earlier). Also, need to add pod identity webhooks targeting other
+cloud providers and change the annotations to more generic naming.
+
+**HyperShift changes**: Include Cloud Credential Operator with "Token" mode (see above). Allows for processing of 
+CredentialsRequest objects added by operators.
+
+**OperatorHub and Console changes**: Allow for import of additional, well-known, bundle annotations which will result 
+in ENV variables on the Subscription object. Show operator is enabled for Token-based use based on Console determining 
+cluster is running in token-based authentication enabled cluster by reading the `.spec.serviceAccountIssuer` from the Authentication CR, `.spec.platformSpec.type` from the Infrastructure CR,`.spec.credentialsMode` from the CloudCredentials CR  AND
+bundle has token-based auth enabled annotation. This info should be written to a new field on Subscription objects.
+The Subscription fields will allow UX for the information needed by CCO or webhook, for input of the cloud credentials. 
+Generate a manual ack for operator upgrade when cloud resources are involved. This can be determined by parsing the 
+Subscription objects. Ensures admin signs off that cloud resources are provisioned and working as intended before an 
+operator upgrade.
+
+**Operator team/ Operator SDK changes**: Follow new guidelines for allowing for the operator to work on token auth 
+enabled cluster. New guidelines would include the following to use CCO Token mode:
+
+- changing CRD to add CredentialRequest references and putting those referenced
+- CredentialRequests into a defined directory in the bundle
+- Add a bundle annotation claiming token-based authentication support
+- add the projected ServiceAccount volume to deployment;
+- handle the Secret themselves (like today’s CCO Mint mode, read and use the credentials in the Secret whose name they 
+  know from the CredentialsRequest and is in the same Namespace)
+- Guidance to operator authors to structure controller code to alert on missing cloud credentials, not crash.
+
+SDK to support this new template. SDK to validate and in particular: alert on any permission changes between operator
+versions.
+
+### Workflow Description
+
+Making a layered operator ready to work on an STS enabled cluster will involve the following steps:
+
+For the operator author team:
+- Add CredentialRequests to the bundle, known location;
+- Add "role-arn"-type fields to the CredentialRequests
```

Comment:
Also it doesn't seem like something the operator cares about. OLM and CCO are the ones connecting the dots here.

Reply:
Nevermind, as you said, the operator is responsible for mutating the CredentialsRequest.

Reply:
I guess to put my question another way:

the operator may be responsible for creating a CredReq that includes the ARN(we need to decide if the operator creates that directly, or olm creates it as part of installing the operator...i haven't seen that made clear, nor am i sure how it works today for non-STS clusters?)  But first it has to get the ARN from somewhere, and that's the responsibility of the console+OLM.

Reply:
@jharrington22 could you answer here? I'll incorporate what you say in the comment to clarify the enhancement text.

Reply:
Today the operator or a user will create the credentials request. Having OLM apply the credentials request is not unreasonable and would have an impact on classic OSD where Red Hat provides the AWS account. 

I think for this proposal we should separate the creation of CredentialsRequests by OLM to be considered later. What component creates the CredentialsRequest wont impact how CCO handles it.

Reply:
for purposes of this EP i mostly just want us to feel confident about what the operator will have to do.

this, for example, is not viable today:
```
Add CredentialRequests to the bundle, known location;
```

because olm won't do anything w/ a credreq in a bundle.

and if the operator is creating the credreq itself, then the operator needs to know to:
1) consume the ARN from somewhere (presumably env on the deployment that is there as  result of the subscription)
2) inject that ARN into the credreq it is creating, properly.

i agree this is largely orthogonal to "enhance the CCO to generate Secrets for STS-enabled clusters", so if the intent is to scope the EP only to the CCO changes, then we can just remove all this discussion about how the CredReq is created and how the Secret is consumed because it's not relevant to the CCO changes.

but i think if we want to be confident that the CCO changes will be useful to the target audience(operator authors), it behooves us to clearly and explicitly define the entire workflow including the pieces the operator will need to implement(and as noted on the call today, there might be 2 or more ways an operator could implement things)

Reply:
@bparees It seems like this thread hasn't been addressed fully? I'd like to push people to start looking at re-reviewing and moving towards approving this EP. Are there changes you can suggest that would resolve this?

Reply:
i think it is resolved now since it states that the operator will be creating the credreq, whereas before it said the credreq would be static in the bundle.